pkg/frontend/querymiddleware/astmapper/sharding.go

```diff
@@ -528,11 +528,12 @@ func shardVectorSelector(curshard, shards int, selector *parser.VectorSelector)
 		return nil, err
 	}
 	return &parser.VectorSelector{
-		Name:           selector.Name,
-		Offset:         selector.Offset,
-		OriginalOffset: selector.OriginalOffset,
-		Timestamp:      copyTimestamp(selector.Timestamp),
-		StartOrEnd:     selector.StartOrEnd,
+		Name:                 selector.Name,
+		Offset:               selector.Offset,
+		OriginalOffset:       selector.OriginalOffset,
+		Timestamp:            copyTimestamp(selector.Timestamp),
+		SkipHistogramBuckets: selector.SkipHistogramBuckets,
```

Comment:
From how the skip histogram bucket works, you cannot really break things by omitting it. Omitting it would just mean that we copy buckets as well in the histogram iterator - so more data.

Conversely, adding it would only break if in the sharded query we started to look at buckets suddenly, which would mean adding a function call like `histogram_quantile`, `histogram_fraction`, `histogram_stddev`, `histogram_stdvar`. But we never do, we just modify the query to add sum(), which is safe.

Reply:
But this question does point out that we need a generic way to handle optimizations on the AST so we can just re-run the optimization without having to worry about this on a case by case basis @beorn7

Reply:
No better insights to share from my side, unfortunately. It sounds to me we have to check optimization on a case by case basis to see how they interact with sharded query evaluation.